core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -960,19 +960,27 @@ private[spark] class TaskSetManager(
       assert (null != failureReason)
       taskSetExcludelistHelperOpt.foreach(_.updateExcludedForFailedTask(
         info.host, info.executorId, index, failureReason))
-      numFailures(index) += 1
-      if (numFailures(index) >= maxTaskFailures) {
-        logError("Task %d in stage %s failed %d times; aborting job".format(
-          index, taskSet.id, maxTaskFailures))
-        abort("Task %d in stage %s failed %d times, most recent failure: %s\nDriver stacktrace:"
-          .format(index, taskSet.id, maxTaskFailures, failureReason), failureException)
-        return
+      if (!successful(index)) {
+        numFailures(index) += 1
+        if (numFailures(index) >= maxTaskFailures && !successful(index)) {
```

Comment:
we don't need to check the condition `!successful(index)` here.

Reply:
I not sure that if there is a thread-safe problem.  `handleFailedTask` could be invoked by both `TaskSetManager.executorLost` and `TaskSchedulerImpl.handleFailedTask`, and `TaskSchedulerImpl.handleSuccessfulTask` and `TaskSchedulerImpl.markPartitionCompleted` would set `successful(index)` to true.

Reply:
There should be no thread-safe problem as all calls are protected by `TaskSchedulerImpl.synchronized`.

Reply:
Agree with both @jiangxb1987 and @Ngone51, please remove `!successful(index)`.

Reply:
removed, thanks all.